### ISSUE TITLE
client lr_scheduler does not need to be one of DeepSpeed's VALID_LR_SCHEDULES

### ISSUE THREAD

**Issue by linxihui** [huggingface/accelerate#1095]
I think this checking is not necessary and wrong. DeepSpeed supports any of `torch.optim.lr_scheduler`. Indeed, when specific in the DeepSpeed config, it will first look up the DeepSpeed `lr_schedulers` (`VALID_LR_SCHEDULES`), and if not found, used those inside `torch.optim.lr_scheduler`.    This checking will prevent us from using pytorch's lr_scheduler with DeepSpeed.

https://github.com/huggingface/accelerate/blob/ea36b7dcebd5292afba328291a2ec6361dbcfb9a/src/accelerate/accelerator.py#L1285-L1286

Also, I think the loop here is incorrect. You shouldn't need to do 8 steps when 8 GPUs.  Unless you define a step as 'micro-step' instead of global step. 

https://github.com/huggingface/accelerate/blob/ea36b7dcebd5292afba328291a2ec6361dbcfb9a/src/accelerate/scheduler.py#L74-L80

**Comment by muellerzr** [comment]
Re; the second, that is correct. Given how accelerate splits the batches of data it's considered a "micro step" https://huggingface.co/docs/accelerate/concept_guides/performance#learning-rates

cc @pacman100 for the DeepSpeed part

**Comment by linxihui** [comment]
@muellerzr  Thanks for replying. I never use without DeepSpeed.  How does it interact with `grad_accum_steps`? Said, if I increase from 2 GPUs to 4GPUs, from `grad_accum_step`=2 to 1, same `per_gpu_batch_size`.  What will happen?

With DeepSpeed, it is definitely wrong (that's what current case, as the lr_scheduler will not be passed to `deepspeed.initialize`, and fall back to use Accelerator scheduler, which step 8 times for 8GPUs). In DeepSpeed, only one optimizer step will be taken in grad_accum boundary.

**Comment by pacman100** [comment]
Hello @linxihui, it is not wrong or unnecessary. If the user has specified a PyTorch Scheduler, it won't be given to the `deepspeed.initialize` and it will be wrapped separately by Accelerate to create `AcceleratedScheduler` object which the user can control separately in their training loop. See, https://github.com/huggingface/accelerate/blob/ea36b7dcebd5292afba328291a2ec6361dbcfb9a/src/accelerate/accelerator.py#L1292-L1298. DeepSpeed docs specify below guidelines for lr scheduler

> if the schedule is supposed to execute at every training step, then the user can pass the scheduler to deepspeed.initialize when initializing the DeepSpeed engine and let DeepSpeed manage it for update or save/restore.
> 
> if the schedule is supposed to execute at any other interval (e.g., training epochs), then the user should NOT pass the scheduler to DeepSpeed during initialization and must manage it explicitly.

When user specifies custom scheduler in training loop, it would be incorrect to assume that it will be executed every training step, as such we wrap it in `AcceleratedScheduler` and let user control it in their training loop.

If the scheduler is one of DeepSpeed's `VALID_LR_SCHEDULES ` or a `DummyScheduler` with its config specified in the DeepSpeed Config file, then only `DeepSpeedEngine` object will have the Scheduler whose update it will control. Please refer to the documentation section `Important code changes when using DeepSpeed Config File` when specifying `optimizer` or `scheduler` in deepspeed config file: https://huggingface.co/docs/accelerate/usage_guides/deepspeed#deepspeed-config-file

**Comment by pacman100** [comment]
@linxihui, with respect to scehduler step updates, let us walk through an example to make sure we are on same page and to confirm that things are working as expected:
1. I specified below optimizer and scheduler config in my ds config file:
```json
"optimizer": {
        "type": "AdamW",
        "params": {
            "lr": "auto",
            "weight_decay": "auto"
        }
    },
    "scheduler": {
        "type": "WarmupDecayLR",
        "params": {
            "warmup_min_lr": "auto",
            "warmup_max_lr": "auto",
            "warmup_num_steps": "auto",
            "total_num_steps": "auto"
        }
    }
``` 
2. Because of this, in my training loop, I will specify `DummyOptim` and `DummyScheduler` as shown below:
```
# Creates Dummy Optimizer if `optimizer` was spcified in the config file else creates Adam Optimizer
 optimizer_cls = (
     torch.optim.AdamW
     if accelerator.state.deepspeed_plugin is None
     or "optimizer" not in accelerator.state.deepspeed_plugin.deepspeed_config
     else DummyOptim
 )
 optimizer = optimizer_cls(optimizer_grouped_parameters, lr=args.learning_rate)

 # Creates Dummy Scheduler if `scheduler` was spcified in the config file else creates `args.lr_scheduler_type` Scheduler
 if (
     accelerator.state.deepspeed_plugin is None
     or "scheduler" not in accelerator.state.deepspeed_plugin.deepspeed_config
 ):
     lr_scheduler = get_scheduler(
         name=args.lr_scheduler_type,
         optimizer=optimizer,
         num_warmup_steps=args.num_warmup_steps,
         num_training_steps=args.max_train_steps,
     )
 else:
     lr_scheduler = DummyScheduler(
         optimizer, total_num_steps=args.max_train_steps, warmup_num_steps=args.num_warmup_steps
     )
```

3. Let:
gradient_accumulation_steps=1
num_processes or num_gpus=8, 
len(train_dataloader) = 100
num_epochs = 10
lr = 5e-5
num_warmup_steps=0

`max_train_steps =  (num_epochs*dataloader_len)/(Gradient_accumulation_steps)` = (100*10)/(1) = 1000. 
But it needs to be divided by `num_processes` to account for data parallel. This is done while replacing the `auto` value of ds config file as shown below:

https://github.com/huggingface/accelerate/blob/ea36b7dcebd5292afba328291a2ec6361dbcfb9a/src/accelerate/accelerator.py#L1270-L1275

So, `scheduler.params.total_num_steps=125` (1000/8). Now, DeepSpeed `model.step()` with call `scheduler.step()` once for first step to reduce lr to 4.96e-5 (i.e., 5e-5 * (1-1/125)) after step 1.

However, for `AcceleratedScheduler` doesn't reduce the `max_train_steps` to 125 and keeps it as 1000. So, after first call to `accelerated_scheduler.step()` it will result into lr being 4.96e-5 (i.e., 5e-5 * (1-8/1000)) . 

So, both result in same lr post first step and so on.

**Comment by linxihui** [comment]
@pacman100  thanks so much for the examples! super detail. I agree with your point. 

As a user, who use the high-level interface of  `model, opt, lr_scheduler = accelerator.prepare(model, opt, lr_shceduler)` with DeepSpeed backend, I'd expect the `lr_scheduler`  be called by DeepSpeed Engine, i.e., once every global batch, just like passing `lr_scheduler` to `deepspeed.initialize`, regardless if it is from `torch.optim.lr_scheduler` or DeepSpeed's `VALID_LR_SCHEDULES`. 

But now instead, my `lr_scheduler` is wrapped by `AcceleratedScheduler` with different executing scheduler.  I think there should be at lease some warning so users know they need to account for #GPUs used, or adding some flag to force it.

**Comment by aldopareja** [comment]
Came to this same problem. gradient accumulation and learning rate warm up was not working when using huggingface's get_scheduler function.

To fix it I had to inject Deepspeed so the scheduler is passed to deepspeed.initialize and you get the same behavior you get when using deepspeed under accelerate.

```python
def patch_target_module(
    to_patch: str,
    replace_with: Any,
):
    to_patch = to_patch.split(".")
    assert len(to_patch) > 1, "must have an object to patch"

    to_patch, obj_name_to_patch = to_patch[:-1], to_patch[-1]
    to_patch = ".".join(to_patch)
    source = importlib.import_module(to_patch)
    setattr(source, obj_name_to_patch, replace_with)
    
from deepspeed.runtime.lr_schedules import VALID_LR_SCHEDULES

new_lr_schedules = deepcopy(VALID_LR_SCHEDULES)
new_lr_schedules.append("LambdaLR")

patch_target_module(
    "deepspeed.runtime.lr_schedules.VALID_LR_SCHEDULES",
    new_lr_schedules,
)

accelerator = Accelerator(
    deepspeed_plugging=...,
)

lr_scheduler = get_scheduler(
    name=args.lr_scheduler,
    optimizer=optimizer,
    num_warmup_steps=args.num_warmup_steps,
    num_training_steps=args.num_epochs * len(train_loader) // grad_accum,
)
model, optimizer, _, lr_scheduler = accelerator.prepare(
    model,
    optimizer,
    train_loader,
    lr_scheduler,
)
```
